src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchShareLogDAO.scala

```diff
@@ -0,0 +1,133 @@
+import org.elasticsearch.action.admin.indices.create.{CreateIndexRequest, CreateIndexRequestBuilder, CreateIndexResponse}
+import org.elasticsearch.action.admin.indices.exists.indices.{IndicesExistsRequest, IndicesExistsRequestBuilder, IndicesExistsResponse}
+import org.elasticsearch.action.get.{GetRequest, GetRequestBuilder, GetResponse}
+import org.elasticsearch.action.index.{IndexRequest, IndexRequestBuilder, IndexResponse}
+import org.elasticsearch.action.search.{SearchRequest, SearchRequestBuilder, SearchResponse}
+import org.elasticsearch.action.support.WriteRequest.RefreshPolicy
+import org.elasticsearch.client.transport.TransportClient
+import org.elasticsearch.common.xcontent.XContentType
+import org.elasticsearch.index.query.QueryBuilder
+import org.elasticsearch.index.query.QueryBuilders._
+import spray.json._
+import spray.json.DefaultJsonProtocol._
+
+import scala.concurrent.Future
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.util.hashing.MurmurHash3
+import scala.util.{Failure, Success, Try}
+
+trait ShareQueries {
+  def userShares(userId: String): QueryBuilder = termQuery("userId", userId)
+  def userSharesOfType(userId: String, shareType: String) = boolQuery()
+    .must(termQuery("userId", userId))
+    .must(termQuery("shareType", shareType))
+}
+
+/**
+  * DAO that uses ElasticSearch to log and get records of shares.
+  *
+  * @param client      The ElasticSearch client
+  * @param indexName   The name of the target share log index in ElasticSearch
+  * @param refreshMode Using IMMEDIATE - the default - is a performance hit but ensures transactionality
+  *                    of updates across multiple users.
+  *                    TODO would NONE be a better option here?
+  */
+class ElasticSearchShareLogDAO(client: TransportClient, indexName: String, refreshMode: RefreshPolicy = RefreshPolicy.IMMEDIATE)
+  extends ShareLogDAO with ElasticSearchDAOSupport with ShareQueries {
+
+  lazy private final val datatype = "sharelog"
+
+  init // checks for the presence of the index
+
+  /**
+    * Logs a record of a user sharing a workspace, group, or method with a user.
+    *
+    * @param userId     The workbench user id
+    * @param sharee     The email of the user being shared with
+    * @param shareType  The type (workspace, group, or method) see `ShareLog`
+    * @return           The record of the share
+    */
+  override def logShare(userId: String, sharee: String, shareType: String): Share = {
+    val share = Share(userId, sharee, shareType, Instant.now)
+    val id = MurmurHash3.stringHash(userId + sharee + shareType).toString
+    val insert = client
+      .prepareIndex(indexName, datatype, id)
+      .setSource(share.toJson.compactPrint, XContentType.JSON)
+      .setCreate(true) // fail the request if the logged share already exists
+      .setRefreshPolicy(refreshMode)
+
+    executeESRequest[IndexRequest, IndexResponse, IndexRequestBuilder](insert)
+    share
+  }
```

Comment:
might wanna break out the id generation into a reusable method, e.g. 
```
  def generateId(share: Share) = {
    val rawString = Seq(share.userId, share.sharee, share.shareType).mkString("|")
    MurmurHash3.stringHash(rawString).toString
  }
```
I can see needing to use this from other parts of the code sometime.

Reply:
just wondering - why the "|"?

Reply:
Not sure how relevant this will be, but the delimiter allows us to distinguish between (`foo.bar`, `baz`) and (`foo.barb`, `az`)